### PR TITLE
Enable LLD_REPORT_UNDEFINED by default

### PR DESCRIPTION
**Pull request by sbc100**
This makes undefined symbol errors more precise by including the name of the
object that references the undefined symbol.

Its also paves the way (in my mind anyway) for finally fixing reverse dependencies
in a salable way.  See #15982.  That PR uses an alternative script for the pre-processing
of dependencies but also fundamentally relies on processing JS libraries both before
and after linking.

The cost is about 300ms per link operation due to double processing of
the JS libraries, but results are cached so in practice this only
happens the first time a given link command is run (see #18326).

### CODE REVIEW

**Comment by sbc100** [comment]
This change has been behind a flag for a long time now and I think its finally ready to be the default.

File: ChangeLog.md
Comment (kripken) [inline ChangeLog.md:25]:
```suggestion
  undefined symbol. The old behaviour (of allowing all undefined symbols at
```

**Review by kripken — COMMENTED** [review]
Is this the change you mentioned adds 300ms to startup? (that we might want to profile)

File: emcc.py
Comment (kripken) [inline emcc.py:1901]: Also this?

(I'd expect this PR to only flip `LLD_REPORT_UNDEFINED` and nothing else?)

File: test/test_other.py
Comment (kripken) [inline test/test_other.py:11485]: Why is there an extra input file here? I guess to scan the symbols for error reporting?

**Review by kripken — COMMENTED** [review]
How does speed look now?

**Comment by sbc100** [comment]
> How does speed look now?

Now that we have the cache amortized cost of "JS symbol generation" in 17ms our of total of 500ms needed to link "hello world".  So that is 3.6%.. and that is fixed cost so for real programs it will be a much much smaller percentage

File: emcc.py
Comment (sbc100) [reply emcc.py:1901]: This PR doesn't just enable LLD_REPORT_UNDEFINED but also fixes all the remaining issues with it.   If you like I can make a separate PR called "fix remaining LLD_REPORT_UNDEFINED issues" before landing this change?

File: test/test_other.py
Comment (sbc100) [reply test/test_other.py:11485]: Actually this wasn't needed in the end. Reverted.

File: emcc.py
Comment (sbc100) [reply emcc.py:1872]: Ditto, this wasn't needed in the final version.  Sorry about that.

**Comment by kripken** [comment]
Speed sounds good!

File: emcc.py
Comment (kripken) [reply emcc.py:1900]: Yes, if that's not hard to do it sounds better? Or, writing up what those fixes are in this PR.

Comment (sbc100) [reply emcc.py:1900]: Done.  This PR is now just flipping the default and updating some test expectations.

Comment (kripken) [reply emcc.py:1900]: I still see a diff on this line? lgtm aside from that.

Comment (sbc100) [reply emcc.py:1900]: Ah yes, I fixed LLD_REPORT_UNDEFINED so we no longer need to force main to be defined.  That would break existing builds.

With this change we can turn on LLD_REPORT_UNDEFINED without requiring main function by default.

One day we can try to switch the IGNORE_MISSING_MAIN default to false, but for now we keep at at true so that this change is as NFC as possible.

Comment (sbc100) [reply emcc.py:1900]: (The comment below is no longer true which is why I removed it)

Comment (sbc100) [reply emcc.py:1900]: The required for `IGNORE_MISSING_MAIN` to be disabled in `LLD_REPORT_UNDEFINED` mode was removed in #17175, BTW, not this change.

**Review by kripken — APPROVED** [review]
Got it, thanks, lgtm.

**Comment by curiousdannii** [comment]
I tried updating to 3.1.28 and I now get errors for undefined symbols when an EM_JS is defined in a library.

```
wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_to_lower_case_uni
wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_to_lower_case_uni
wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_to_upper_case_uni
wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_to_upper_case_uni
wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_to_title_case_uni
wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_to_title_case_uni
wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_canon_decompose_uni
wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_canon_decompose_uni
wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_canon_normalize_uni
wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_canon_normalize_uni
```

The EM_JS functions are defined [in a library](https://github.com/curiousdannii/emglken/blob/939991800f91fa3e50b71a72e0c874b1c610fa6b/src/cgunicod.c#L175). If the functions are then used in a executable then it makes these errors.

I'm not sure why, but [another EM_ASYNC_JS function](https://github.com/curiousdannii/emglken/blob/939991800f91fa3e50b71a72e0c874b1c610fa6b/src/getc.c#L16) doesn't cause any errors. Maybe because it's only used by the same file?

**Comment by sbc100** [comment]
> Sorry to raise an error just after you concluded there were no issues stuck_out_tongue But you know, sometimes it takes a while to try updating dependencies.
> 
> I tried updating to 3.1.28+ and I now get errors for undefined symbols when an EM_JS function is defined in a library.
> 
> ```
> wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_to_lower_case_uni
> wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_to_lower_case_uni
> wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_to_upper_case_uni
> wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_to_upper_case_uni
> wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_to_title_case_uni
> wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_to_title_case_uni
> wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_canon_decompose_uni
> wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_canon_decompose_uni
> wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_canon_normalize_uni
> wasm-ld: error: libremglk.a(gi_dispa.c.o): undefined symbol: glk_buffer_canon_normalize_uni
> ```
> 

I think the reason for this is that the EM_JS macro basically expands to a declaration of the EM_JS function as external with an `__attribute__((import_name))` attached to it.  

However, when you use it outside in other compilation units they don't see that declaration.  Actually that sounds odd now that I say its since those attributes should probably attached to the final symbol.

In any case, I think the soltution that others have uses it to only call the EM_JS function within the file where they are defined, and then define small (difined) wrappers for them for use outside the compilation unit.

e.g. declare `glk_buffer_to_title_case_uni_js` as the EM_JS function then define a local `glk_buffer_to_title_case_uni` which then just calls it.

(I will investigate further to see if we can actually this fix with you having to change any code).

**Comment by curiousdannii** [comment]
Sure, that makes sense. If you can find a work around that would be great, otherwise that is a simple solution.

If you can't find a workaround, then the docs should also get an update saying that it's only safe to call a EM_JS etc function from the same file.

**Pull request merged**
